RIPS/rip-9995.md

```diff
@@ -0,0 +1,199 @@
+```
+uint256 version, uint256 action, bytes32 txHash, bytes transaction
+```
+
+The frame must return a bytes array that is interpreted as:
+
+```solidity
+abi.encode(context, MAGIC_VALUE_PAYMASTER, validUntil, validAfter)
+```
+
+#### Sender Execution
+As the beginning of actual transaction execution, this call is made to the `Sender` contract directly,
+meaning it will be executed starting at the first byte of code section 0, and `pc` set to 0.
+
+Inputs to the `Sender` contract are not defined by the protocol and are controlled by the `callData` parameter.
+
+#### Paymaster post-transaction frame
+
+Inputs to the `Paymaster` post-transaction are defined by the protocol as a packed encoding of the following data:
+```
+uint256 version, uint256 action, uint256 actualGasCost, bool success, bytes context
+```
+
+## Rationale
+
+### ABI encoding for system frames' input and output data
+
+Using an encoding format for data provided by the protocol itself does represent an abstraction levels violation,
+it is a relatively safe one any alternative solution would require some trade-offs.
+
+The validation section of a Smart Contract Account code needs to have full access to the majority of transaction
+details in order to be able to make an informed decision about either accepting or rejecting the transaction.
+
+A small subset of this data is available with the existing opcodes, however creating an opcode for every transaction
+parameter is not feasible.
+
+Allowing wallets to specify their own encoding for this data is also not feasible as Smart Contract Accounts must
+avoid any ambiguity about the meaning of the received data.
+
+Using a different encoding will not benefit from the fact that ABI decoding logic is already built into
+Solidity programming language and ships with most of deployed smart contracts.
+
+### Allowing regular code execution flows to access `kind_abstraction` sections
+
+The code in a `kind_abstraction` does not have any difference
+There is absolutely no difference between `kind_abstraction (0x05)` and `kind_code (0x02)` sections outside the
+context of Native Account Abstraction transactions.
+
+The `kind_abstraction` specifies an alternative starting point for a system frame execution, but after that
+code from different sections can be accessed as well.
+
+There seems to be no reason to isolate `kind_abstraction` code section in any way.
+
+### Using `action` identifier instead of different kind markers
+
+As an alternative, we could define multiple kind markers, for example `kind_validate (0x05)`, `kind_validate_paymaster`,
+`kind_deploy_account` and `kind_paymaster_post_transaction`.
+
+However, this approach would fill the kind markers namespace with very particular types of code sections with little
+practical difference between them. This approach could also complicate creation of multi-role smart contracts,
+for example a contract sharing the `Paymaster` and `Deployer` functionality.
+
```

Comment:
I think it's the opposite. A multi-role contract will be clearly visible as such, having different sections instead of one section with a nonstandard multiplexing function for actions.  Also slightly cheaper to execute (no need to multiplex).

Reply:
yes, it clears the difference. but I actually think we should avoid multi-role contracts.
Apaymaster contract is not an account - it is a contract.
An SCA is more akin to "EOA", which is triggered by an off-chain owner.